### DISCUSSION TITLE
Miryoku Customisation Documentation

### DISCUSSION THREAD

**Discussion by manna-harbour** [manna-harbour/miryoku#85] [Announcements]
**This is a draft of the new Miryoku customisation documentation.**

- For questions and suggestions on the documentation please comment here.
- For issues with the implementation, see https://github.com/manna-harbour/miryoku/issues/94.
- Personalisation by selection of included alternative layout and mapping options is covered by existing documentation.
- This document applies to Miryoku QMK, Miryoku ZMK, Miryoku KMonad, and Miryoku KMK, unless noted otherwise.
- If you'd like to share your customisations with the community see https://github.com/manna-harbour/miryoku/discussions/8.
  - To share your `custom_config.h`:
    - [Create a discussion in Show and Tell and include `custom_config.h` as part of the title](https://github.com/manna-harbour/miryoku/discussions/new?category=show-and-tell&title=Sample%20custom_config.h%20by%20NAME).
    - Add links to your `custom_config.h`. Use live links, permalinks to the file, permalinks to lines or regions, or a gist.
    - Include the following summary template and check the boxes as appropriate:
```
## Summary
- Miryoku Implementations
  - [ ] Miryoku QMK
  - [ ] Miryoku ZMK
  - [ ] Miryoku KMonad
  - [ ] Miryoku KMK
- Features
  - Layers
    - [ ] Substitute Layer
    - [ ] Add Layer
    - [ ] Remove Layer
    - [ ] Change Layer Display Name
  - Change Mapping
    - [ ] All Layers
    - [ ] Per Layer
```

## custom_config.h

`custom_config.h` can be used to substitute, add, and remove layers, and to substitute mappings for all layers or per layer, in a single file and without causing merge conflicts with updates.

For complete examples, see [example `custom_config.h` shared by members of the community](https://github.com/manna-harbour/miryoku/discussions/categories/show-and-tell?discussions_q=category%3A%22Show+and+tell%22+custom_config.h).

In Miryoku QMK, `custom_config.h` is https://github.com/manna-harbour/miryoku_qmk/blob/miryoku/users/manna-harbour_miryoku/custom_config.h. In Miryoku ZMK, `custom_config.h` is https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/custom_config.h and is described at https://github.com/manna-harbour/miryoku_zmk#config-file. In Miryoku KMK, `custom_config.h` is https://github.com/manna-harbour/miryoku_kmk/blob/main/src/custom_config.h. 

### Layers

Layer data is defined in `miryoku_babel/miryoku_layer_alternatives.h` and selected in `miryoku_babel/miryoku_layer_selection.h` at build time depending on build options.

#### Substitute Layer

Copy the layer definition to be modified from `miryoku_layer_alternatives.h` into `custom_config.h`, change the macro name to match the name for the layer used in `miryoku_layer_selection.h`, and edit the definition as appropriate.

E.g. to replace apostrophe with semicolon on Base when using `MIRYOKU_ALPHAS=QWERTY`, copy the `#define MIRYOKU_ALTERNATIVES_BASE_QWERTY` block (5 lines) from `miryoku_layer_alternatives.h` into `custom_config.h`, and change `MIRYOKU_ALTERNATIVES_BASE_QWERTY` to `MIRYOKU_LAYER_BASE`. For Miryoku QMK replace `KC_QUOT` with `KC_SCLN`. For Miryoku ZMK replace `SQT` with `SEMI`.  For Miryoku KMonad replace `U_QUOT` with `;`. For Miryoku KMK replace `KC.QUOT` with `KC.SCLN`.

#### Add Layer

Define the layer data in `custom_config.h` in the same way as when substituting a layer, but use a new layer name in the macro name. Available layers are listed in `miryoku_babel/miryoku_layer_list.h`. Copy the `#define MIRYOKU_LAYER_LIST` block (11 lines) into `custom_config.h` and add an entry for the new layer. Note the layer names in this list (and in layer data macro names) do not have the `U_` prefix.  The name in quotes is for display. If not also changing the per layer mapping for this layer, add a default for `MIRYOKU_LAYERMAPPING_` plus layer name to `MIRYOKU_MAPPING`. E.g. for a new layer `GAME`, add `#define MIRYOKU_LAYERMAPPING_GAME MIRYOKU_MAPPING`.

For Miryoku ZMK and Miryoku KMK, `miryoku_layer_list.h` also contains a `#define` for each layer.  These lines must also be copied, a new entry added, and all entries renumbered.

#### Remove Layer

Copy the layer list in the same way as when adding a layer, then remove the entry for the layer to be removed.

Layers are referenced from other layers, so to avoid having to modify other layers, add a `#define` to remap the missing layer onto another layer. E.g. When removing the Fun layer, add `#define U_FUN U_NUM` to remap all references to the Num layer instead.

For Miryoku ZMK and Miryoku KMK, remove the `#define` for the removed layer and renumber all entries as when adding a layer, but keep the additional `#define` that remaps the layer.

#### Miryoku QMK

To change the mapping for all layers, copy the mapping macro definition to `custom_config.h`, change `LAYOUT_miryoku` to `MIRYOKU_MAPPING`, and edit as appropriate.

E.g. to map the left hand top row outer pinkie column key on `crkbd` to Escape, copy the contents of `layouts/community/split_3x6_3/manna-harbour_miryoku/config.h` to `custom_config.h`, change `LAYOUT_miryoku` to `MIRYOKU_MAPPING`, and change the first `XXX` in the macro definition to `KC_ESC`.

To change the mapping per layer, do as above, but change `LAYOUT_miryoku` to `MIRYOKU_LAYERMAPPING_` plus layer name. E.g. for Base, change to  `MIRYOKU_LAYERMAPPING_BASE`.

As mappings need to be specific to the keyboard, everything should be wrapped in `#if defined` / `#endif` conditional on the automatically defined `KEYBOARD_` plus keyboard name macro. E.g. to make the mapping specific to builds for `crkbd` use `#if defined (KEYBOARD_crkbd)` / `#endif`.

#### Miryoku ZMK

To change the mapping for all layers, copy the mapping macro definition (everything inside the `#if defined` / `#endif` conditional on the `MIRYOKU_LAYOUTMAPPING_` plus layout name macro)  to `custom_config.h`, and edit as appropriate.

E.g. to map the left hand top row outer pinkie column key on `corne` to Escape, copy everything inside the `#if !defined (MIRYOKU_LAYOUTMAPPING_CORNE)` / `#endif` block from `miryoku/mapping/42/corne.h` to `custom_config.h`, and change the first `XXX` in the macro definition to `&kp ESC`.

To change the mapping per layer, do as above, but change `MIRYOKU_LAYOUTMAPPING_` plus layout name to `MIRYOKU_LAYERMAPPING_` plus layer name. E.g. for `corne` and Base, change `MIRYOKU_LAYOUTMAPPING_CORNE` to  `MIRYOKU_LAYERMAPPING_BASE`.

As per layer mappings need to be specific to the keyboard, they should be wrapped in `#if defined` / `#endif` conditional on the automatically defined `MIRYOKU_KEYBOARD_` plus keyboard name macro. E.g. for `corne` use `#if defined (MIRYOKU_KEYBOARD_CORNE)` / `#endif`. Note that the automatic macro is only available for workflow builds and will need to be manually added to the top of `custom_config.h` for local builds, e.g. for `corne` use `#define MIRYOKU_KEYBOARD_CORNE`.

#### Miryoku KMonad

Edit the `defsrc` section. Also see https://github.com/manna-harbour/miryoku/issues/90.

#### Miryoku KMK

To change the mapping for all layers, copy the mapping macro definition (everything inside the `#if defined` / `#endif` conditional on the `MIRYOKU_LAYOUTMAPPING_` plus layout name macro)  to `custom_config.h`, and edit as appropriate.

E.g. to map the left hand top row outer pinkie column key on `CRKBD` to Escape, copy everything inside the `#if !defined (MIRYOKU_LAYOUTMAPPING_CRKBD)` / `#endif` block from `miryoku/mapping/42/crkbd.h` to `custom_config.h`, and change the first `XXX` in the macro definition to `KC.ESC`.

To change the mapping per layer, do as above, but change `MIRYOKU_LAYOUTMAPPING_` plus layout name to `MIRYOKU_LAYERMAPPING_` plus layer name. E.g. for `CRKBD` and Base, change `MIRYOKU_LAYOUTMAPPING_CRKBD` to  `MIRYOKU_LAYERMAPPING_BASE`.

As per layer mappings need to be specific to the keyboard, they should be wrapped in `#if defined` / `#endif` conditional on the automatically defined `KMK_KEYBOARD_` plus keyboard name macro. E.g. for `CRKBD` use `#if defined (KMK_KEYBOARD_CRKBD)` / `#endif`. 

## Miryoku Babel

Miryoku Babel can be used to modify, add, and remove layers for all implementations from a single source.  To modify layers, edit the corresponding source tables. To add layers, add new source tables. To add or remove layers, edit the `layers` table, the `layer-body` source block, and the `miryoku_layer_alternatives.h` source blocks (other than the first few lines, the contents of the `miryoku_layer_alternatives.h` source blocks can be copied between implementations). 

## Custom Hardware Support

Only the hardware support enabled at the keyboard level is included by default.

### Miryoku QMK

For hardware support or other custom code, add to `rules.mk`, `config.h`, and `keymap.c` in the keyboard's `keymaps/manna-harbour_miryoku/` directory. For keyboards supported via community layouts, first change to a keyboard keymap, e.g. for `ckrbd`, create `keyboards/crkbd/keymaps/manna-harbour_miryoku/` and copy `config.h` and `keymap.c` from `layouts/community/split_3x6_3/manna-harbour_miryoku/` to `keyboards/crkbd/keymaps/manna-harbour_miryoku/`.

To copy from another keymap, copy the appropriate lines from `rules.mk`, `config.h`, and `keymap.c` into the corresponding files in the keyboard's `keymaps/manna-harbour_miryoku/` directory. E.g. to copy RGB and OLED support from the default `crkbd` keymap, copy `keyboards/crkbd/keymaps/default/rules.mk` to `keyboards/crkbd/keymaps/harbour_miryoku/rules.mk`, append `keyboards/crkbd/keymaps/default/config.h` to `keyboards/crkbd/keymaps/harbour_miryoku/config.h`, and copy `keyboards/crkbd/keymaps/default/keymap.c` to `keyboards/crkbd/keymaps/harbour_miryoku/keymap.c` and remove the `keymaps` array (i.e. [this block](https://github.com/qmk/qmk_firmware/blob/master/keyboards/crkbd/keymaps/default/keymap.c#L22-L71)).

### Miryoku ZMK

See [Kconfig Configuration](https://github.com/manna-harbour/miryoku_zmk/#kconfig-configuration).

### Miryoku KMK

See [Installation](https://github.com/manna-harbour/miryoku_kmk#installation).

**Comment by agisga** [comment]
Thank you, this looks great! Is there also an "official" way to modify/add combos in addition to the available thumb combos? I have been modifying `manna-harbour_miryoku.c` so far to achieve that (in QMK), but can one of the `custom*` files be used now instead?

**Reply by fourstepper** [reply]
combos would indeed be pretty cool, but not sure if it would fit the scope of the miryoku project?

**Reply by agisga** [reply]
I find it very convenient to use combos for custom macros (for example, opening and navigating applications I use often, typing out often-used character sequences, etc.). This could be used on top of the standard miryoku keymap, without making any changes to the positioning of any existing keys (just like rebinding extra keys on a larger keyboard without changing any keys used by miryoku). I don't feel like such a use-case is "out-of-scope" (because it would not change the underlying standard miryoku but just add something on top of it)?

**Reply by manna-harbour** [reply]
I have avoided using combos for the main layout. Heavy use is ergonomically questionable, and can require certain switches and caps. Also, leaving them out makes them available for things outside the main layout, such as key emulation combos.

But I do recommend adding combos as a way to optimise for your workflow without needing to modify the layers, as they can be easily changed at any time without affecting the rest of the layout. E.g. frequently used symbols, or macros for multi-symbol operators, particularly if they include a mixture of symbols from different layers.  So I would like to make that easier, and then add that suggestion to the documentation.

I've created a new issue that covers this: https://github.com/manna-harbour/miryoku/issues/94. I'd like to keep this discussion for only the documentation, but please comment on that issue if you'd like to discuss adding extra features to the implementation.

**Comment by zthreee** [comment]
I have problem with this documentation when it comes to chancing not-used keys. 

Looking at -> Mapping - Miryoku QMK: 

I have Lily58, so I went to keybards/lily58/keymaps/manna-harbour_miryoku/config.h and copied the whole config.h which is: 

```
#pragma once

#define XXX KC_NO

#define LAYOUT_miryoku(\
     K00, K01, K02, K03, K04,                K05, K06, K07, K08, K09,\
     K10, K11, K12, K13, K14,                K15, K16, K17, K18, K19,\
     K20, K21, K22, K23, K24,                K25, K26, K27, K28, K29,\
     N30, N31, K32, K33, K34,                K35, K36, K37, N38, N39\
)\
LAYOUT(\
XXX, XXX, XXX, XXX, XXX, XXX,                XXX, XXX, XXX, XXX, XXX, XXX,\
XXX, K00, K01, K02, K03, K04,                K05, K06, K07, K08, K09, XXX,\
XXX, K10, K11, K12, K13, K14,                K15, K16, K17, K18, K19, XXX,\
XXX, K20, K21, K22, K23, K24, XXX,      XXX, K25, K26, K27, K28, K29, XXX,\
               XXX, K32, K33, K34,      K35, K36, K37, XXX\
)
```

Added this into custom_config.h in users/manna-harbour_miryoku/custom_config.h

Changed XXX keys and name of the layout from `LAYOUT_miryoku` to `MIRYOKU_MAPPING`

```
#pragma once

#define MIRYOKU_MAPPING(\
     K00, K01, K02, K03, K04,                K05, K06, K07, K08, K09,\
     K10, K11, K12, K13, K14,                K15, K16, K17, K18, K19,\
     K20, K21, K22, K23, K24,                K25, K26, K27, K28, K29,\
     N30, N31, K32, K33, K34,                K35, K36, K37, N38, N39\
)\
LAYOUT(\
K_ESC, KC_1, KC_2, KC_3, KC_4, KC_5,                KC_6, KC_7, KC_8, KC_9, KC_0, KC_GRAVE,\
KC_TAB, K00, K01, K02, K03, K04,                K05, K06, K07, K08, K09, KC_BACKSPACE,\
KC_LGUI, K10, K11, K12, K13, K14,                K15, K16, K17, K18, K19, KC_QUOT,\
KC_LSFT, K20, K21, K22, K23, K24, KC_NO,      KC_NO, K25, K26, K27, K28, K29, KC_RSFT,\
               KC_NO, K32, K33, K34,      K35, K36, K37, KC_NO\
)
```

Now when using qmk compile -c -kb lily58 -km manna-harbour_miryoku they keys arent mapped properly - what exactly I am doing wrong here? 

___ 

Edit: Checked one option and I modified the config.h in lily58/keymaps/manna-harbour_miryoku/config.h to 

```
#pragma once

#define XXX KC_NO

#define LAYOUT_miryoku(\
     K00, K01, K02, K03, K04,                K05, K06, K07, K08, K09,\
     K10, K11, K12, K13, K14,                K15, K16, K17, K18, K19,\
     K20, K21, K22, K23, K24,                K25, K26, K27, K28, K29,\
     N30, N31, K32, K33, K34,                K35, K36, K37, N38, N39\
)\
LAYOUT(\
    KC_ESC,   KC_1, KC_2, KC_3,  KC_4, KC_5,                KC_6, KC_7, KC_8, KC_9, KC_0, KC_GRAVE,\
    KC_TAB,   K00,  K01,  K02,   K03,  K04,                 K05, K06, K07, K08, K09, KC_BACKSPACE,\
    KC_LGUI,  K10,  K11,  K12,   K13,  K14,                 K15, K16, K17, K18, K19, KC_QUOT,\
    KC_LSFT,  K20,  K21,  K22,   K23,  K24, KC_NO,      KC_NO, K25, K26, K27, K28, K29, KC_RSFT,\
                        KC_NO,   K32,  K33,   K34,            K35, K36, K37, KC_NO\
)
```

And now it works. 

But I would like to get it working on the first way of doing this - to copy to custom_config.h 
orr.. the custom_config.h should be in the same folder as the original config.h in lily context?

**Reply by manna-harbour** [reply]
Thanks for pointing that out. For Miryoku QMK, this feature currently requires the Miryoku QMK development branch. I've just added a note to the document. I'll be making a PR soon, so it will end up in QMK master after that.

**Comment by Hugin0** [comment]
When it says to redefine a layer (when removing it) to another layer, could I say define it to any valid keycode, i.e. replace the mouse layer with KC_LSFT?

**Reply by manna-harbour** [reply]
No, the layer name macro is used in a layer-tap, so it needs to expand to a layer number. If you'd like to use Shift for the hold action you'd substitute a custom Base layer as described here, and replace the Mouse layer tap with a Shift mod tap.

**Comment by hanifn** [comment]
> For hardware support or other custom code, add to rules.mk, config.h, and keymap.c in the keyboard's keymaps/manna-harbour_miryoku/ directory. 

So I've been doing it wrong [here](https://github.com/hanifn/qmk_firmware/blob/bf66f55c48f534b3fcebf1a99749c80eeb3887db/users/manna-harbour_miryoku/manna-harbour_miryoku.c#L109-L603) and all that code should live in `keyboards/boardsource/lulu/keymaps/manna-harbour_miryoku/keymap.c`?

**Reply by manna-harbour** [reply]
At some point there will be a `custom_keymap.c` or similar, and that will be the best place for custom code that you want to use with more than one keyboard / community layout. See https://github.com/manna-harbour/miryoku/issues/94.

If you have hardware specific code that only applies to one keyboard then the keyboard `keymap.c` would be best.

For now, what you have is good. You already have `#ifdef OLED_ENABLE` so it's fine on boards with no OLED. You might also want to add a conditional on display size if you want to support other displays. The main point of `custom_config.h` is to make customisation more manageable (avoid merge conflicts and keep changes together) but as your changes are all at the end of the file you should be pretty safe.

**Comment by partap** [comment]
I'm looking for a way to dynamically swap between Win/Linux and Mac layouts, i.e.

- Swap clipboard keycodes between using Ctrl and Cmd (like MIRYOKU_CLIPBOARD=MAC, but dynamic)
- Swap Ctrl and Cmd modifier positions in general (in QMK, I use CG_SWAP/CG_NORM for that, but not sure how with ZMK)

Just starting out with zmk (and miryoku_zmk) on a new Corne keyboard...
On my "old" Iris, I've been using QMK and a custom keymap that's been moving closer and closer to a full reimplementation of miryaku.  My main issue with the Iris is the teeny tiny memory size (and no bluetooth)

**Reply by manna-harbour** [reply]
This Discussion is to discuss the above customisation documentation only. For customisation support other than issues with this documentation, see https://github.com/manna-harbour/miryoku#discussions-and-support. I've created a new discussion for your issue at https://github.com/manna-harbour/miryoku/discussions/192.

**Comment by yingyang11** [comment]
Can you also add columns (e.g. 2 columns for 42 key corne) too if you just added them to custom_config.h, or do I need to specify that there are 12 columns somewhere else too?

**Reply by manna-harbour** [reply]
In general you can't just add new keys since they wouldn't correspond to any keys in the keyboard definition.

But in the case of 6 column Corne, the Miryoku mapping for 5 and 6 column Cornes is the same so that you can easily map the extra keys. This works because a 5 column Corne is just a 6 column Corne without some switches installed.

**Reply by mrclayman** [reply]
Thanks for the response, @manna-harbour , but I am not sure I fully get that. I have just put together a 42-key Corne while I still have a 36-key Chocofi that I have been using until now, both of which use the Corne keymap. Is there a way for me to accommodate both keebs with the existing macro? The `MIRYOKU_LAYOUTMAPPING_CORNE` macro currently has 40 arguments, but 4 are kind of dummy because they refer to the 4 missing keys in the bottommost row. I suppose I could modify the above macro to accommodate more arguments, but that does not seem to be the correct way to do it.

**Reply by manna-harbour** [reply]
@mrclayman You need to replace the unused keys in the macro body: https://github.com/manna-harbour/miryoku_zmk/blob/559aa4beae75cb3206ab411b7da2adb9665c6896/miryoku/mapping/42/corne.h#L14-L17. You don't need to modify the arguments. See the Mapping section above. Those keys won't be used on the 36 key board but can be used on the 42 key board, with the same mapping.

**Reply by mrclayman** [reply]
Hi @manna-harbour , thanks for the response. If I understand what you're saying correctly, that means that I must put in specific bindings right into the corne.h? Will that not make the outermost column on the Corne have the same function for all layers? I don't think I can customize the key functions like this, but I am probably missing something else here. Thanks again for your time.
